text/0114-resolved-ts-for-large-transactions.md

```diff
@@ -0,0 +1,164 @@
+# Resolved-ts for Large Transactions
+
+Author: @ekexium
+
+Tracking issue: https://github.com/tikv/tikv/issues/17459
+
+## 1. Background
+
+This RFC is a variation and extension of @zhangjinpeng87's [Large Transactions Don't Block Watermark](https://github.com/pingcap/tiflow/blob/master/docs/design/2024-01-22-ticdc-large-txn-not-block-wm.md). They aim to solve the same problem.
+
+Resolved-ts is a mechanism that provides a temporal guarantee. It is defined as follows: Once a resolved-ts T is observed by any component of the system, it is guaranteed that no new commit record with a commit-ts less than or equal to T will be subsequently produced or observed by any component of the system.
+
+In current TiKV(v8.3), large transactions can block resolve-ts from advancing, because it is calculated as `min(pd-tso, min(lock.ts))`, which is actually a more stringent constraint than its original definition. A lock from a pipelined txn can live several hours. This will make services dependent on resolved-ts unavailable.
+
+## 2. Goals
+
+- Current phase objectives:
+  - Prevent pipelined transactions from impeding resolved-ts advancement.
+
+- Long-term goal:
+  - Minimize the overhead of resolved-ts maintenance.
+  - Maximize resolved-ts freshness.
+  - Achieve uninterrupted resolved-ts progression, addressing all potential blocking factors beyond large transactions.
+
+## 3. Proposed Design
+
+### 3.1 Key Idea
+
+**Proposed Change:** Utilize `lock.min_commit_ts` instead of `lock.start_ts` for resolved-ts calculation.
+
+Rationale:
+1. Resolved-ts definition is independent of LOCK CF.
+2. Current use of `lock.start_ts` is based on the invariant: `lock.start_ts` < `lock.commit_ts`.
+3. A valid resolved-ts doesn't necessitate the absence of locks with smaller start_ts, as long as their future commit_ts are guaranteed to be larger.
+
+Advantages of `lock.min_commit_ts`:
+1. Satisfies a similar invariant: `lock.min_commit_ts` <= its `commit_ts`
+2. Unlike the constant `lock.start_ts`, `lock.min_commit_ts` can be dynamically increased.
+
+### 3.2 Maintenance of resolved-ts
+
+Key objective: Maximize all TiKV nodes' awareness of large pipelined transactions during their lifetime.
+
+#### 3.2.1 Coordinator
+
+**Proposed Change:** 
+- TTL manager fetches latest TSO as min_commit_ts candidate.
```

Comment:
Is the `TTL manager` the worker in tidb which updates the TTL of primary key for each transaction?

Reply:
Yes, because we want to piggyback the `min_commit_ts` in the heartbeat message.